### PR TITLE
New Resource : `azurerm_cognitive_deployment`

### PR DESCRIPTION
**Pull request by sinbai**
I would like to explain that for property`scale_settings` , it should contain two properties [capacity](https://github.com/Azure/azure-rest-api-specs/blob/83cde4fa7d6c820f574b9aec973fd9597385e312/specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2022-10-01/cognitiveservices.json#L2790) and [scaleType](https://github.com/Azure/azure-rest-api-specs/blob/83cde4fa7d6c820f574b9aec973fd9597385e312/specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2022-10-01/cognitiveservices.json#L2778). After confirming with the service team, the `capacity` is not support at this time (GA date 2022-12-15). For convenience of future expansion, I reserved the `scale_settings` property, although it only has one property  `scaleType`.

Overview doc link : https://learn.microsoft.com/en-us/azure/cognitive-services/openai/overview

### CODE REVIEW

**Review by katbyte — CHANGES_REQUESTED** [review]
We have test failues: 
image

**Review by katbyte — CHANGES_REQUESTED** [review]
We still have test failures, we may need to pick different regions for the open AI tests to pass in the teamcity .kt file?

```
------- Stdout: -------
=== RUN   TestAccCognitiveDeploymentSequential
=== RUN   TestAccCognitiveDeploymentSequential/deployment
=== RUN   TestAccCognitiveDeploymentSequential/deployment/basic
    testcase.go:117: Step 1/2 error: Error running apply: exit status 1
        
        Error: creating Account (Subscription: "*******"
        Resource Group Name: "acctest-rg-221211230757865856"
        Account Name: "acctest-ca-221211230757865856"): cognitiveservicesaccounts.CognitiveServicesAccountsClient#AccountsCreate: Failure sending request: StatusCode=0 -- Original Error: Code="SpecialFeatureOrQuotaIdRequired" Message="The subscription does not have QuotaId/Feature required by SKU 'S0' from kind 'OpenAI' ."
        
          with azurerm_cognitive_account.test,
          on terraform_plugin_test.tf line 23, in resource "azurerm_cognitive_account" "test":
          23: resource "azurerm_cognitive_account" "test" {
        
=== RUN   TestAccCognitiveDeploymentSequential/deployment/requiresImport
    testcase.go:117: Step 1/2 error: Error running apply: exit status 1
        
        Error: creating Account (Subscription: "*******"
        Resource Group Name: "acctest-rg-221211230856046784"
        Account Name: "acctest-ca-221211230856046784"): cognitiveservicesaccounts.CognitiveServicesAccountsClient#AccountsCreate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidApiSetId" Message="The account type 'OpenAI' is either invalid or unavailable in given region."
        
          with azurerm_cognitive_account.test,
          on terraform_plugin_test.tf line 23, in resource "azurerm_cognitive_account" "test":
          23: resource "azurerm_cognitive_account" "test" {
        
=== RUN   TestAccCognitiveDeploymentSequential/deployment/complete
    testcase.go:117: Step 1/2 error: Error running apply: exit status 1
        
        Error: creating Account (Subscription: "*******"
        Resource Group Name: "acctest-rg-221211231057693582"
        Account Name: "acctest-ca-221211231057693582"): cognitiveservicesaccounts.CognitiveServicesAccountsClient#AccountsCreate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidApiSetId" Message="The account type 'OpenAI' is either invalid or unavailable in given region."
        
          with azurerm_cognitive_account.test,
          on terraform_plugin_test.tf line 22, in resource "azurerm_cognitive_account" "test":
          22: resource "azurerm_cognitive_account" "test" {
        
--- FAIL: TestAccCognitiveDeploymentSequential (241.10s)
    --- FAIL: TestAccCognitiveDeploymentSequential/deployment (241.10s)
        --- FAIL: TestAccCognitiveDeploymentSequential/deployment/basic (58.18s)
        --- FAIL: TestAccCognitiveDeploymentSequential/deployment/requiresImport (121.65s)
        --- FAIL: TestAccCognitiveDeploymentSequential/deployment/complete (61.27s)
FAIL
```

**Comment by sinbai** [comment]
> We still have test failures, we may need to pick different regions for the open AI tests to pass in the teamcity .kt file?
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccCognitiveDeploymentSequential
> === RUN   TestAccCognitiveDeploymentSequential/deployment
> === RUN   TestAccCognitiveDeploymentSequential/deployment/basic
>     testcase.go:117: Step 1/2 error: Error running apply: exit status 1
>         
>         Error: creating Account (Subscription: "*******"
>         Resource Group Name: "acctest-rg-221211230757865856"
>         Account Name: "acctest-ca-221211230757865856"): cognitiveservicesaccounts.CognitiveServicesAccountsClient#AccountsCreate: Failure sending request: StatusCode=0 -- Original Error: Code="SpecialFeatureOrQuotaIdRequired" Message="The subscription does not have QuotaId/Feature required by SKU 'S0' from kind 'OpenAI' ."
>         
>           with azurerm_cognitive_account.test,
>           on terraform_plugin_test.tf line 23, in resource "azurerm_cognitive_account" "test":
>           23: resource "azurerm_cognitive_account" "test" {
>         
> === RUN   TestAccCognitiveDeploymentSequential/deployment/requiresImport
>     testcase.go:117: Step 1/2 error: Error running apply: exit status 1
>         
>         Error: creating Account (Subscription: "*******"
>         Resource Group Name: "acctest-rg-221211230856046784"
>         Account Name: "acctest-ca-221211230856046784"): cognitiveservicesaccounts.CognitiveServicesAccountsClient#AccountsCreate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidApiSetId" Message="The account type 'OpenAI' is either invalid or unavailable in given region."
>         
>           with azurerm_cognitive_account.test,
>           on terraform_plugin_test.tf line 23, in resource "azurerm_cognitive_account" "test":
>           23: resource "azurerm_cognitive_account" "test" {
>         
> === RUN   TestAccCognitiveDeploymentSequential/deployment/complete
>     testcase.go:117: Step 1/2 error: Error running apply: exit status 1
>         
>         Error: creating Account (Subscription: "*******"
>         Resource Group Name: "acctest-rg-221211231057693582"
>         Account Name: "acctest-ca-221211231057693582"): cognitiveservicesaccounts.CognitiveServicesAccountsClient#AccountsCreate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidApiSetId" Message="The account type 'OpenAI' is either invalid or unavailable in given region."
>         
>           with azurerm_cognitive_account.test,
>           on terraform_plugin_test.tf line 22, in resource "azurerm_cognitive_account" "test":
>           22: resource "azurerm_cognitive_account" "test" {
>         
> --- FAIL: TestAccCognitiveDeploymentSequential (241.10s)
>     --- FAIL: TestAccCognitiveDeploymentSequential/deployment (241.10s)
>         --- FAIL: TestAccCognitiveDeploymentSequential/deployment/basic (58.18s)
>         --- FAIL: TestAccCognitiveDeploymentSequential/deployment/requiresImport (121.65s)
>         --- FAIL: TestAccCognitiveDeploymentSequential/deployment/complete (61.27s)
> FAIL
> ```

Hi @katbyte , thanks for your review. Per the [doc](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/quotas-limits), only 2 OpenAI resources could be created per region, I updated the tests to run sequentially. Also, per the [doc ](https://azure.microsoft.com/en-us/pricing/details/cognitive-services/openai-service/#overview), OpenAI is not supported in all regions. We need to set env.ARM_PROVIDER_DYNAMIC_TEST = true to ensure tests run in supported `westeurope`. After updating the above tow points, the tests passed as follows:
image

**Review by katbyte — CHANGES_REQUESTED** [review]
@sinbai - you need to update the settings.kt file to lock the tests to regions that support it like this:
```
// Custom Providers is only available in certain locations
        "customproviders" to testConfiguration(locationOverride = LocationConfiguration("eastus", "westus2", "westeurope", true)),
```

**Comment by sinbai** [comment]
> @sinbai - you need to update the settings.kt file to lock the tests to regions that support it like this:
> 
> ```
> // Custom Providers is only available in certain locations
>         "customproviders" to testConfiguration(locationOverride = LocationConfiguration("eastus", "westus2", "westeurope", true)),
> ```
@katbyte thanks for letting me know this. I have updated the settings.kt file. But the configuration of teamcity setting is not changed, could you please help me to see if there is something missing to update? Details are as follows:

> 1. **settings.kt file:**

image
 

>  2.  **PR 19526 teamcity setting:**

image
image

**Comment by katbyte** [comment]
@sinbai - changes wont take affect until we merge this PR, until this the new locations just have to be passed in manually intot he configuration

File: internal/services/cognitive/cognitive_deployment_resource.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						ForceNew: true,
+						ValidateFunc: validation.StringInSlice([]string{
+							string(deployments.DeploymentScaleTypeStandard),
+						}, false),
+					},
+				},
+			},
+		},
```
Comment (katbyte) [inline internal/services/cognitive/cognitive_deployment_resource.go:114]: settings is redundant in tf? and scale is in the block name?
```suggestion
		"scale": {
			Type:     pluginsdk.TypeList,
			Required: true,
			ForceNew: true,
			MaxItems: 1,
			Elem: &pluginsdk.Resource{
				Schema: map[string]*pluginsdk.Schema{
					"type": {
						Type:     pluginsdk.TypeString,
						Required: true,
						ForceNew: true,
						ValidateFunc: validation.StringInSlice([]string{
							string(deployments.DeploymentScaleTypeStandard),
						}, false),
					},
				},
			},
		},
```

**Review by katbyte — APPROVED** [review]
Thanks, aside from one comment on schema  this LGTM

File: internal/services/cognitive/cognitive_deployment_resource.go (hunk 1 above)
Comment (sinbai) [reply internal/services/cognitive/cognitive_deployment_resource.go]: @katbyte thanks for your feedback. The suggestion has been fixed. Could you please take another look?

**Comment by sinbai** [comment]
> 

@katbyte  got it. Thanks for letting me know.

File: internal/services/cognitive/cognitive_deployment_resource.go (hunk 1 above)
Comment (sinbai) [reply internal/services/cognitive/cognitive_deployment_resource.go]: @katbyte thanks for your feedback. The suggestion has been fixed. Could you please take another look?

**Comment by sinbai** [comment]
@katbyte this PR has been merged for about a month, it has not been released yet. Any issue blocked it? May I ask the ETA about release this PR?

**Pull request merged**